promql/query_logger.go

```diff
@@ -0,0 +1,217 @@
+	"os"
+	"strings"
+	"syscall"
+	"time"
+)
+
+type QueryLogger struct {
+	mmapedFile   []byte
+	getNextIndex chan int
+	lastIndex    chan int
+	logger       log.Logger
+	Ok           bool
+}
+
+const (
+	entrySize int = 200
+)
+
+func parseBrokenJson(brokenJson []byte, logger log.Logger) (bool, string) {
+	queries := strings.TrimSpace(string(brokenJson))
+	queries = strings.Trim(queries, "\x00")
+	queries = queries[:len(queries)-1] + "]"
+
+	// conditional because of implementation detail: len() = 1 implies file consisted of a single char: '['
+	if len(queries) == 1 {
+		return false, "[]"
+	}
+
+	var out bytes.Buffer
+	err := json.Indent(&out, []byte(queries), "", "\t")
+
+	if err != nil {
+		level.Warn(logger).Log("msg", "Error parsing json", "err", err)
+		return false, "[]"
+	}
+
+	return true, out.String()
+}
+
+func logUnfinishedQueries(filename string, filesize int, logger log.Logger) {
+	if _, err := os.Stat(filename); err == nil {
+		fd, err := os.Open(filename)
+		if err != nil {
+			level.Error(logger).Log("msg", "Failed to open query log file", "err", err)
+			return
+		}
+
+		brokenJson := make([]byte, filesize)
+		_, err = fd.Read(brokenJson)
+		if err != nil {
+			level.Error(logger).Log("msg", "Failed to read query log file", "err", err)
+			return
+		}
+
+		queriesExist, queries := parseBrokenJson(brokenJson, logger)
+		if !queriesExist {
+			return
+		}
+		level.Info(logger).Log("msg", "These queries didn't finish in prometheus' last run:")
+		fmt.Println(queries)
+	}
+}
+
+func getMMapedFile(filename string, filesize int, logger log.Logger) (error, []byte) {
+
+	fd, err := os.OpenFile(filename, os.O_CREATE|os.O_RDWR|os.O_TRUNC, 0644)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error opening query log file", "file", filename, "err", err)
+		return err, []byte{}
+	}
+
+	if err := fd.Truncate(int64(filesize)); err != nil {
+		level.Error(logger).Log("msg", "Error creating file of specified size", "file", filename, "Attemped size", filesize, "err", err)
+		return err, []byte{}
+	}
+
+	fileAsBytes, err := syscall.Mmap(int(fd.Fd()), 0, filesize, syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_SHARED)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error mmapping file", "file", filename, "err", err)
+		return err, []byte{}
+	}
+
+	return err, fileAsBytes
+}
+
+func NewQueryLogger(filename string, filesize int, logger log.Logger) (queryLogger QueryLogger) {
+	defer func() {
+		if r := recover(); r != nil {
+			queryLogger = QueryLogger{}
+		}
+	}()
+
+	logUnfinishedQueries(filename, filesize, logger)
+
+	if filesize < 1+entrySize {
+		level.Error(logger).Log("msg", "Filesize must be greater than entrysize.", "entrysize", entrySize)
+		return QueryLogger{}
+	}
+
+	err, fileAsBytes := getMMapedFile(filename, filesize, logger)
+	if err != nil {
+		return QueryLogger{}
+	}
+
+	copy(fileAsBytes, "[")
+	queryLogger = QueryLogger{
+		mmapedFile:   fileAsBytes,
+		getNextIndex: make(chan int, (1<<16)-1),
```

Comment:
Why 32k? We know the maximum number of running queries from flags.

Reply:
Oh didn't think of that. I'll fix it.